src/ripple/basics/SlabAllocator.h

```diff
@@ -30,67 +30,176 @@
+    //           '---------------'
+    //
+    // Note that the order of slots within the free list at any given time
+    // is random, as it depends on the pattern and timing of memory access
+    // of the application.
+    struct alignas(Type) FreeItem
+    {
+        FreeItem* next;
+
+        FreeItem(FreeItem* n = nullptr) noexcept : next(n)
+        {
+        }
+
+        ~FreeItem() noexcept = default;
+    };
+
+    static_assert(sizeof(Type) >= sizeof(FreeItem));
+
     /** A block of memory that is owned by a slab allocator */
     struct SlabBlock
     {
-        // A mutex to protect the freelist for this block:
+        // A mutex to protect this block:
         std::mutex m_;
 
         // A linked list of appropriately sized free buffers:
-        std::uint8_t* l_ = nullptr;
+        FreeItem* list_ = nullptr;
 
-        // The next memory block
+        // The next block:
         SlabBlock* next_;
 
         // The underlying memory block:
-        std::uint8_t const* const p_ = nullptr;
+        std::uint8_t* data_ = nullptr;
+
+        // The number of items available in this block
+        std::size_t avail_{0};
+
+        explicit SlabBlock(SlabBlock* next) : next_(next)
+        {
+        }
 
-        // The extent of the underlying memory block:
-        std::size_t const size_;
+        // We don't release the allocated memory here: slabs are only
+        // destroyed when the process is shutting down, and we cannot
+        // guarantee that this isn't called before the destruction of
+        // all the objects that are allocated in the buffer we have.
+        ~SlabBlock() = default;
 
-        SlabBlock(
-            SlabBlock* next,
+        std::uint8_t*
+        assign_and_allocate(
             std::uint8_t* data,
             std::size_t size,
-            std::size_t item)
-            : next_(next), p_(data), size_(size)
+            std::size_t item,
+            std::size_t count)
         {
-            // We don't need to grab the mutex here, since we're the only
-            // ones with access at this moment.
+            assert(
+                reinterpret_cast<std::uintptr_t>(data) %
+                    alignof(decltype(data_)) ==
+                0);
+
+            assert(count == (size / item));
+
+            assert(data != nullptr);
+
+            std::lock_guard _(m_);
+
+            assert(
+                (avail_ == 0 && list_ == nullptr) ||
+                (avail_ != 0 && list_ != nullptr));
 
-            while (data + item <= p_ + size_)
+            // It's possible that this block has an underlying buffer. If
+            // it does, and it has space available, we use it.
+            if (data_ != nullptr)
             {
-                // Use memcpy to avoid unaligned UB
-                // (will optimize to equivalent code)
-                std::memcpy(data, &l_, sizeof(std::uint8_t*));
-                l_ = data;
-                data += item;
+                auto ret = list_;
```

Comment:
Technically I think we need to run a destructor on `*list_` or this is UB. I don't think it actually matters in practice, but we may as well be technically correct here, and it's an easy change.

Reply:
The destructor is trivial (see [`std::is_trivially_destructible`](https://en.cppreference.com/w/cpp/types/is_destructible)) which, if I understand things correctly, means that the destructor does nothing and so not to not explicitly invoke it  is fine.

With that said, I'm perfectly fine with adding a call to `std::destroy_at(ret)` here and it might make sense to eliminate any confusion, even if it's technically correct, which is the best kind of correct.

Note that there is a somewhat analogous situation in `SHAMapItem` (which I introduced) that it may make sense to also address by removing the `if constexpr` check and calling `std::destroy_at` unconditionally:

https://github.com/XRPLF/rippled/blob/c3acbce82dc5db8765cfc9dfeae36954cec0312f/src/ripple/shamap/SHAMapItem.h#L146-L150

Reply:
I think it's UB if we don't run the destructor even if it's trivailly destructable. Yes, I agree it won't matter in practice, but I think we should do it (unless someone can correct me and show it's not UB).

Reply:
I unfortunately agree that we have to run the destructor even if trivial.  The optimizers have gotten too scary when it comes to UB.

Reply:
Interesting. I'll update the code to run the destructor here and update the code in `SHAMapItem`. Thanks @HowardHinnant!